staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go

```diff
@@ -92,30 +105,49 @@ func Compile(s *schema.Structural, isResourceRoot bool) ([]CompilationResult, er
 	// compResults is the return value which saves a list of compilation results in the same order as x-kubernetes-validations rules.
 	compResults := make([]CompilationResult, len(celRules))
 	for i, rule := range celRules {
-		var compilationResult CompilationResult
-		if len(strings.TrimSpace(rule.Rule)) == 0 {
-			// include a compilation result, but leave both program and error nil per documented return semantics of this
-			// function
-		} else {
-			ast, issues := env.Compile(rule.Rule)
-			if issues != nil {
-				compilationResult.Error = &Error{ErrorTypeInvalid, "compilation failed: " + issues.String()}
-			} else if !proto.Equal(ast.ResultType(), decls.Bool) {
-				compilationResult.Error = &Error{ErrorTypeInvalid, "cel expression must evaluate to a bool"}
-			} else {
-				prog, err := env.Program(ast)
-				if err != nil {
-					compilationResult.Error = &Error{ErrorTypeInvalid, "program instantiation failed: " + err.Error()}
-				} else {
-					compilationResult.Program = prog
-				}
+		compResults[i] = compileRule(rule, env)
+	}
+
+	return compResults, nil
+}
+
+func compileRule(rule apiextensions.ValidationRule, env *cel.Env) (compilationResult CompilationResult) {
+	if len(strings.TrimSpace(rule.Rule)) == 0 {
+		// include a compilation result, but leave both program and error nil per documented return semantics of this
+		// function
+		return
+	}
+	ast, issues := env.Compile(rule.Rule)
+	if issues != nil {
+		compilationResult.Error = &Error{ErrorTypeInvalid, "compilation failed: " + issues.String()}
+		return
+	}
+	if !proto.Equal(ast.ResultType(), decls.Bool) {
+		compilationResult.Error = &Error{ErrorTypeInvalid, "cel expression must evaluate to a bool"}
+		return
+	}
+
+	if checkedExpr, err := cel.AstToCheckedExpr(ast); err != nil {
+		// should be impossible since env.Compile returned no issues
+		compilationResult.Error = &Error{ErrorTypeInternal, "unexpected compilation error: " + err.Error()}
+		return
+	} else {
```

Comment:
nit, why else after the return?

Reply:
Thanks, I overlooked this in the latest revision while changing the function signature from `f(*CompilationResult)` to `f() CompilationResult`. Fixed.